### PR TITLE
[e2e] Add customizable `flutter_driver` adaptor

### PR DESCRIPTION
**Pull request by CareF**
## Description

Add a customizable `flutter_driver` adaptor, allowing changing timeout time, flag to enable timeline tracing. 
A use case is [here](https://github.com/CareF/measure_overhead) where I'm using the timeline to estimate overhead of `E2E` and `flutter_driver` (E2E is better!) 

## Related Issues

Not a full implementation but can be considered a work around for flutter/flutter#58789

## Test

It seems that we don't yet have unit test for E2E. I changed the related example. 

## Checklist

Before you create this PR confirm that it meets all requirements listed below by checking the relevant checkboxes (`[x]`). This will ensure a smooth and quick review process.

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] My PR includes unit or integration tests for *all* changed/updated/fixed behaviors (See [Contributor Guide]).
- [x] All existing and new tests are passing.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] The analyzer (`flutter analyze`) does not report any problems on my PR.
- [x] I read and followed the [Flutter Style Guide].
- [x] The title of the PR starts with the name of the plugin surrounded by square brackets, e.g. [shared_preferences]
- [x] I updated pubspec.yaml with an appropriate new version according to the [pub versioning philosophy].
- [x] I updated CHANGELOG.md to add a description of the change.
- [x] I signed the [CLA].
- [x] I am willing to follow-up on review comments in a timely manner.

## Breaking Change

Does your PR require plugin users to manually update their apps to accommodate your change?

- [ ] Yes, this is a breaking change (please indicate a breaking change in CHANGELOG.md and increment major revision).
- [x] No, this is *not* a breaking change.

[issue database]: https://github.com/flutter/flutter/issues
[Contributor Guide]: https://github.com/flutter/plugins/blob/master/CONTRIBUTING.md
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[pub versioning philosophy]: https://www.dartlang.org/tools/pub/versioning
[CLA]: https://cla.developers.google.com/

### CODE REVIEW

File: packages/e2e/example/test_driver/example_e2e_test.dart [hunk 1]
```diff
@@ -2,4 +2,4 @@ import 'dart:async';
 
 import 'package:e2e/e2e_driver.dart' as e2e;
 
-Future<void> main() async => e2e.main();
+Future<void> main() async => e2e.e2eDriver();
```
Comment (dnfield) [inline packages/e2e/example/test_driver/example_e2e_test.dart:5]: Why this change?

File: packages/e2e/lib/e2e_driver.dart [hunk 2]
```diff
@@ -4,15 +4,49 @@ import 'dart:io';
...
+///
+/// Future<void> main() async => e2e.e2eDriver();
+///
+/// ```
+Future<void> e2eDriver({
+  Duration timeout = const Duration(minutes: 1),
+  String testName,
+  bool traceTimeline = false,
+}) async {
+  print(testName);
```
Comment (dnfield) [inline packages/e2e/lib/e2e_driver.dart:27]: Remove print

File: packages/e2e/lib/e2e_driver.dart [hunk 3]
```diff
@@ -4,15 +4,49 @@ import 'dart:io';
...
+  Future<void> runner() async {
+    jsonResult = await driver.requestData(null, timeout: timeout);
+  }
+
+  if (traceTimeline) {
+    timeline = await driver.traceAction(runner);
+  } else {
+    await runner();
+  }
   final e2e.Response response = e2e.Response.fromJson(jsonResult);
```
Comment (dnfield) [inline packages/e2e/lib/e2e_driver.dart:40]: Won't this now be null sometimes and throw?

File: packages/e2e/lib/e2e_driver.dart [hunk 4]
```diff
@@ -4,15 +4,49 @@ import 'dart:io';
...
+/// in the app's `test_driver` directory:
+///
+/// ```dart
+/// import 'dart:async';
+///
+/// import 'package:e2e/e2e_driver.dart' as e2e;
+///
+/// Future<void> main() async => e2e.e2eDriver();
+///
+/// ```
```
Comment (dnfield) [inline packages/e2e/lib/e2e_driver.dart:21]: The doc string should explain the parameters and how/why to use them.

File: packages/e2e/example/test_driver/example_e2e_test.dart (hunk 1 above)
Comment (CareF) [reply packages/e2e/example/test_driver/example_e2e_test.dart:5]: It's in the example. I'm thinking with the new function `e2e.e2eDriver` is the recommended solution, but `main` is preserved to avoid breaking change.

Comment (dnfield) [reply packages/e2e/example/test_driver/example_e2e_test.dart:5]: But why even change the method here? Why not just add a new test that covers the case you're interested in?

File: packages/e2e/lib/e2e_driver.dart (hunk 3 above)
Comment (CareF) [reply packages/e2e/lib/e2e_driver.dart:40]: Why? `runner` is awaited either directly or in `traceAction`.

File: packages/e2e/example/test_driver/example_e2e_test.dart (hunk 1 above)
Comment (CareF) [reply packages/e2e/example/test_driver/example_e2e_test.dart:5]: I thought these examples are not test cases but demos for users about how to use e2e.

Comment (CareF) [reply packages/e2e/example/test_driver/example_e2e_test.dart:5]: And another thing is, we currently don't have a `test` directory under e2e, I'm not sure how it's unit tested, and this part have to connect to a driver and I don't even know if we are able to run this repo on devicelab.

Comment (dnfield) [reply packages/e2e/example/test_driver/example_e2e_test.dart:5]: We could certainlyl add tests to e2e :)

It's just not clear to me why we're adding this redirecting method and calling it directly instead of just calling main.

File: packages/e2e/lib/e2e_driver.dart (hunk 3 above)
Comment (dnfield) [reply packages/e2e/lib/e2e_driver.dart:48]: Ahh I see now. 

This is a nit, but IMO it would read a bit more clearly if you just directly used start/stop trace if needed, rather than passing a closure to a convenience wrapper.

File: packages/e2e/example/test_driver/example_e2e_test.dart (hunk 1 above)
Comment (CareF) [reply packages/e2e/example/test_driver/example_e2e_test.dart:5]: So `main` is limited for the format of arguments, due to convention and due to if we want `e2e_driver.dart` to be used directly as the driver script. `main` is using the default argument version of `e2eDeiver` but I'm not expecting future users to use `main` anymore.

File: packages/e2e/lib/e2e_driver.dart (hunk 3 above)
Comment (CareF) [reply packages/e2e/lib/e2e_driver.dart:48]: `FlutterDriver` is an abstract class. We have two implementation of it and all override `traceAction`. I would rather still use this method, but add some comments here.

File: packages/e2e/example/test_driver/example_e2e_test.dart (hunk 1 above)
Comment (CareF) [reply packages/e2e/example/test_driver/example_e2e_test.dart:5]: @liyuqian No it shouldn't. On the one hand, It's just an async main, there's nothing it need to await for, on the other hand, return a Future without awaiting is very common in Framework `Future<void> Foo() {return Future<void>.delayed(...)`, where whoever `await` this `Foo` would actually `await` this `Future.delayed`.

Comment (liyuqian) [reply packages/e2e/example/test_driver/example_e2e_test.dart:5]: Oh right, it's ` => e2e.main()` instead of `{ e2e.main(); }`. I think only the latter needs `{ await e2e.main(); }`. Anyway, if we don't `await`, can we remove the `async` and just use `Future<void> main() => e2e.main();`? That way, it's very clear that no `await` is expected there.

File: packages/e2e/lib/e2e_driver.dart [hunk 5]
```diff
@@ -4,15 +4,59 @@ import 'dart:io';
...
 import 'package:flutter_driver/flutter_driver.dart';
 
-Future<void> main() async {
+Future<void> main() async => e2eDriver();
+
+/// Adaptor to run E2E test using `flutter drive`.
+///
+/// `timeout` controls the longest time waited before the test ends.
+/// It is not necessarily the execution time for the test app.
+/// The test can and usually runs for shorter time if the test ends early.
```
Comment (dnfield) [inline packages/e2e/lib/e2e_driver.dart:13]: Maybe just `The test may finish sooner than the timeout`?  I don't think we should say anything about usual.

File: packages/e2e/lib/e2e_driver.dart [hunk 6]
```diff
@@ -4,15 +4,59 @@ import 'dart:io';
...
+/// ```dart
+/// import 'dart:async';
+///
+/// import 'package:e2e/e2e_driver.dart' as e2e;
+///
+/// Future<void> main() async => e2e.e2eDriver();
+///
+/// ```
+Future<void> e2eDriver({
+  Duration timeout = const Duration(minutes: 1),
```
Comment (dnfield) [inline packages/e2e/lib/e2e_driver.dart:32]: Generally we try to avoid timeouts in tests, because they're a source of flakiness.

That said, I'm not sure that 1 minute is long enough here - integration tests can easily take more (for example, they may have to wait on network requests, animations to finish, etc). In the devicelab we use a default of 15 minutes for the entire test run. This should probably be higher by default, maybe something like an hour.

We should also document that these parameters must not be null.

**Comment by dnfield** [comment]
Looking at this a bit more, I think it would be better if we just offered some kind of API to get the timeline at various points, and to help with starting/stopping recording.

I don't think this is fine grained enough - someone may want the timeline for just part of their test, and they may not want to write the timeline to a local file system - they might need the test itself to send the timeline somewhere else.

Can we re-think the design here to handle that kind of use case?

File: packages/e2e/example/test_driver/example_e2e_test.dart (hunk 1 above)
Comment (CareF) [reply packages/e2e/example/test_driver/example_e2e_test.dart:5]: > Oh right, it's ` => e2e.main()` instead of `{ e2e.main(); }`. I think only the latter needs `{ await e2e.main(); }`. Anyway, if we don't `await`, can we remove the `async` and just use `Future<void> main() => e2e.main();`? That way, it's very clear that no `await` is expected there.

Good point! I deleted the `async`.

File: packages/e2e/lib/e2e_driver.dart (hunk 6 above)
Comment (CareF) [reply packages/e2e/lib/e2e_driver.dart:32]: This 1min is what used to be hard coded in `main`. Any suggestion what value I should put here?

**Comment by CareF** [comment]
> Looking at this a bit more, I think it would be better if we just offered some kind of API to get the timeline at various points, and to help with starting/stopping recording.
> 
> I don't think this is fine grained enough - someone may want the timeline for just part of their test, and they may not want to write the timeline to a local file system - they might need the test itself to send the timeline somewhere else.
> 
> Can we re-think the design here to handle that kind of use case?

The hard part is, the test is driven on the device, and the timeline is collected on host. The generation of the timeline is distributed in framework and engine which we have very limited control. To achieve that there will have to be conversation between the device and the host, making the device code depending directly or indirectly on `flutter_driver`, and there will also be delay for this communication.

**Comment by dnfield** [comment]
The timeline is actually connected on the device by the driver extension. It's fine to make it easier to shovel it over to the host to write, but we should expose an easier way to just collect it whether you end up using the driver extension or not.

**Comment by CareF** [comment]
So if the collecting and processing of timeline is completely on device, it makes finer graining but: 
1. It introduces the overhead of parsing json on device; 
2. We need to design a new set of APIs for the timeline collecting, on the back of which we need more codes to connect E2E bindings to vmservice and flutter_driver. 

It should be do-able, but for my project I don't think I should invest more time on it because this feature is not critical for me. It will be great if this PR lands and I can use this feature for my project, otherwise I will put these codes separately in my projects. 

Our (with @liyuqian) original plan is to implement a frame build time statistics API in or out of E2E, but independent of the timeline for performance test, both for the purpose of host independent and for the purpose of using it in release mode.

File: packages/e2e/lib/e2e_driver.dart (hunk 6 above)
Comment (CareF) [reply packages/e2e/lib/e2e_driver.dart:32]: And for the documents I added an `assert(timeout != null);`. I'm expecting it will be obvious since the implementation will be shown together with detailed documents anyway.

**Comment by CareF** [comment]
@dnfield Would you mind letting me know if we finally decide to refuse this PR? I will need to add new features as flutter/flutter#61490 and I don't want to introducing conflicts.

**Comment by dnfield** [comment]
I'd prefer to see a solution for this that does not require a host device to be attached. For example, we should support getting timeline information on Firebase Testlab.

**Comment by CareF** [comment]
> I'd prefer to see a solution for this that does not require a host device to be attached. For example, we should support getting timeline information on Firebase Testlab.

@dnfield I think timeline requires vmservice, which I'm not sure how to have it enabled for an app running completely without a host.

**Comment by liyuqian** [comment]
https://github.com/flutter/flutter/pull/61509/files#diff-73e66281b14c792afcdcd20acb9edd92R5 says that this is blocking that PR. I wonder what's the future plan for this PR?

**Comment by CareF** [comment]
> https://github.com/flutter/flutter/pull/61509/files#diff-73e66281b14c792afcdcd20acb9edd92R5 says that this is blocking that PR. I wonder what's the future plan for this PR?

@liyuqian I didn't update here because @dnfield mentioned: 

> I'd prefer to see a solution for this that does not require a host device to be attached. For example, we should support getting timeline information on Firebase Testlab. 

I think I am not planning on implementing a host-independent timeline collection but #2873 would be an infra for that. I might make sense to modify what's here to make a new PR, with an `e2e_driver` that write the data defined in #2873 to a file, which will be used for flutter/flutter#61509.
